src/waterfall/polarplot.cpp

```diff
@@ -0,0 +1,114 @@
+#include "filemanager.h"
+#include "polarplot.h"
+
+#include <limits>
+
+#include <QtConcurrent>
+#include <QPainter>
+#include <QtMath>
+#include <QVector>
+
+PING_LOGGING_CATEGORY(polarplot, "ping.polarplot")
+
+// Number of samples to display
+uint16_t PolarPlot::_resolution = 400;
+
+PolarPlot::PolarPlot(QQuickItem *parent)
+    :Waterfall(parent)
+    ,_image(2500, 2500, QImage::Format_RGBA8888)
+    ,_painter(nullptr)
+    ,_updateTimer(new QTimer(this))
+{
+    // This is the max depth that ping returns
+    _DCRing.fill({static_cast<float>(_image.height()), 0, 0, 0}, _resolution);
+    setAcceptedMouseButtons(Qt::AllButtons);
+    setAcceptHoverEvents(true);
+    _image.fill(QColor(Qt::transparent));
+
+    connect(_updateTimer, &QTimer::timeout, this, [&] {update();});
+    _updateTimer->setSingleShot(true);
+    _updateTimer->start(50);
+
+    connect(this, &Waterfall::mousePosChanged, this, &PolarPlot::updateMouseColumnData);
+}
+
+void PolarPlot::clear()
+{
+    qCDebug(polarplot) << "Cleaning waterfall and restarting internal variables";
+    _DCRing.fill({static_cast<float>(_image.height()), 0, 0, 0}, _resolution);
+    _image.fill(Qt::transparent);
+}
+
+void PolarPlot::paint(QPainter *painter)
+{
+    static QPixmap pix;
+    if(painter != _painter) {
+        _painter = painter;
+    }
+
+    // http://blog.qt.io/blog/2006/05/13/fast-transformed-pixmapimage-drawing/
+    pix = QPixmap::fromImage(_image, Qt::NoFormatConversion);
+    _painter->drawPixmap(_painter->viewport(), pix, QRect(0, 0, _image.width(), _image.height()));
+}
+
+void PolarPlot::setImage(const QImage &image)
+{
+    _image = image;
+    emit imageChanged();
+    setImplicitWidth(image.width());
+    setImplicitHeight(image.height());
+}
+
+void PolarPlot::draw(const QVector<double>& points, float confidence, float initPoint, float length, float distance)
+{
+    static float currentDrawAngle(0);
+    static const QPoint center(_image.width()/2, _image.height()/2);
+    static const float d2r = M_PI/180.0;
+    static QColor pointColor;
+    static float step;
+    static float angleStep;
+    float actualAngle = currentDrawAngle*d2r;
+
+    // This ring vector will store variables of the last n samples for user access
+    static const float rad2grad = 200/M_PI;
+    _DCRing[int(actualAngle*rad2grad) % 400] = {initPoint, length, confidence, distance};
+
+    for(int i = 1; i < center.x(); i++) {
+        pointColor = valueToRGB(points[i%200]);
+        step = ceil(i*2*d2r)*1.15;
+        for(float u = -0.25; u <= 0.25; u += 1/step) {
+            angleStep = u*d2r+actualAngle - M_PI/2;
+            _image.setPixelColor(center.x() + i*cos(angleStep), center.y() + i*sin(angleStep), pointColor);
```

Comment:
A consideration for optimizing: all of the trigonometry can be done at compile time. In a previous implementation, I identified each possible 'cell' to color as an individual QPath, and all of the QPaths were stored in a two dimensional lookup table accessed by (angle, range):
https://github.com/bluerobotics/ping-viewer/commit/228d66c88854ccd1e8bfa409af3a48cd431a4c80

I'm not sure which implementation is better for supporting dynamic # samples though (ex rectangle container scan).

Reply:
we can leave it unresolved for posterity (:

Reply:
https://github.com/bluerobotics/ping-viewer/issues/515 I just found it, thank you!